File: android/src/main/kotlin/com/pubmatic/sdk/flutter_openwrap_sdk/OpenWrapSDKClient.kt
package com.pubmatic.sdk.flutter_openwrap_sdk

import com.pubmatic.sdk.common.OpenWrapSDK
import com.pubmatic.sdk.common.models.POBApplicationInfo
import com.pubmatic.sdk.common.models.POBLocation
import com.pubmatic.sdk.common.models.POBUserInfo

import com.pubmatic.sdk.flutter_openwrap_sdk.POBUtils.findBy

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

import java.net.URL

/**
 * Wrapper around [OpenWrapSDK] to transfer the dart method calls to OpenWrap SDK.
 */
object OpenWrapSDKClient {

  /**
   * Calls the [OpenWrapSDK] respective method.
   *
   * @param methodName received via methodChannel
   * @param call to fetch the required arguments received from dart side
   * @param result to transfer the result for native side
   */
  @JvmStatic
  fun methodCall(methodName: String, call: MethodCall, result: Result) {
    when (methodName) {
      "setLogLevel" -> {
        call.arguments?.let { argument ->
          val logLevel: OpenWrapSDK.LogLevel? = OpenWrapSDK.LogLevel::getLevel findBy argument
          logLevel?.let { level ->
            OpenWrapSDK.setLogLevel(level)
            result.success(null)
            return
          }
          result.error(
            POBFlutterConstants.OPENWRAP_PLATFORM_EXCEPTION,
            "Error while calling setLogLevel on OpenWrapSDK class.",
            "Cannot set log level as the received log level is null."
          )
        }
      }

      "getVersion" -> {
        result.success(OpenWrapSDK.getVersion())
      }

      "allowLocationAccess" -> {
        call.arguments?.let {
          OpenWrapSDK.allowLocationAccess(it as Boolean)
        }
        result.success(null)
      }

      "setUseInternalBrowser" -> {
        call.arguments?.let {
          OpenWrapSDK.setUseInternalBrowser(it as Boolean)
        }
        result.success(null)
      }

      "setLocation" -> {
        call.argument<Int>("source")?.let { argument ->
          val source = POBLocation.Source.values().getOrElse(argument) { POBLocation.Source.GPS }
          val latitude: Double? = call.argument("latitude")
          val longitude: Double? = call.argument("longitude")
          if (latitude != null && longitude != null) {
            OpenWrapSDK.setLocation(POBLocation(source, latitude, longitude))
            result.success(null)
            return
          }
          result.error(
            POBFlutterConstants.OPENWRAP_PLATFORM_EXCEPTION,
            "Error while calling setLocation on OpenWrapSDK class.",
            "Cannot set location as latitude or longitude is null."
          )
        }
      }

      "setCoppa" -> {
        call.arguments?.let {
          OpenWrapSDK.setCoppa(it as Boolean)
        }
        result.success(null)
      }

      "setSSLEnabled" -> {
        call.arguments?.let {
          OpenWrapSDK.setSSLEnabled(it as Boolean)
        }
        result.success(null)
      }

      "allowAdvertisingId" -> {
        call.arguments?.let {
          OpenWrapSDK.allowAdvertisingId(it as Boolean)
        }
        result.success(null)
      }

      "setApplicationInfo" -> {
        OpenWrapSDK.setApplicationInfo(convertMapToApplicationInfo(call))
        result.success(null)
      }

      "setUserInfo" -> {
        OpenWrapSDK.setUserInfo(convertMapToUserInfo(call))
        result.success(null)
      }

      else -> {
        result.notImplemented()
      }
    }
  }

  @JvmStatic
  private fun convertMapToUserInfo(call: MethodCall): POBUserInfo {
    val userInfo = POBUserInfo()

    call.argument<Int>("birthYear")?.let {
      userInfo.birthYear = it
    }

    call.argument<Int>("gender")?.let { argument ->
      userInfo.setGender(
        POBUserInfo.Gender.values().getOrElse(argument) { POBUserInfo.Gender.MALE })
    }

    call.argument<String>("country")?.let {
      userInfo.setCountry(it)
    }

    call.argument<String>("city")?.let {
      userInfo.setCity(it)
    }

    call.argument<String>("metro")?.let {
      userInfo.setMetro(it)
    }

    call.argument<String>("zip")?.let {
      userInfo.setZip(it)
    }

    call.argument<String>("region")?.let {
      userInfo.setRegion(it)
    }

    call.argument<String>("userKeywords")?.let {
      userInfo.keywords = it
    }

    return userInfo
  }

  @JvmStatic
  private fun convertMapToApplicationInfo(call: MethodCall): POBApplicationInfo {
    val applicationInfo = POBApplicationInfo()

    applicationInfo.domain = call.argument<String>("domain")
    call.argument<String>("storeURL")?.let {
      applicationInfo.storeURL = URL(it)
    }
    call.argument<Boolean>("paid")?.let {
      applicationInfo.setPaid(it)
    }
    applicationInfo.categories = call.argument<String>("categories")
    applicationInfo.keywords = call.argument<String>("appKeywords")

    return applicationInfo
  }
}

File: ios/Classes/OpenWrapSDKClient.swift
import UIKit
import Flutter
import OpenWrapSDK

class OpenWrapSDKClient: NSObject {

/// Method to route all the OpenWrapSDK global API calls
/// - Parameters:
///     - name: method to invoke
///     - call: FlutterMethodCall for argument & other details
///     - result: result block to invoke
    static func invokeMethod(name:String, call: FlutterMethodCall, result: FlutterResult) {

        switch name {
        case "setLogLevel":
            if let logLevel = call.arguments as? Int {
                OpenWrapSDK.setLogLevel(POBSDKLogLevel(rawValue: UInt(logLevel)) ?? .warning)
            }
            result(nil)
        case "getVersion":
            result(OpenWrapSDK.version())
        case "allowLocationAccess":
            if let allow = call.arguments as? Bool {
                OpenWrapSDK.allowLocationAccess(allow)
            }
            result(nil)
        case "setUseInternalBrowser":
            if let use = call.arguments as? Bool {
                OpenWrapSDK.useInternalBrowser(use)
            }
            result(nil)
        case "setLocation":
            if let values = call.arguments as? Dictionary<String, Any>,
               let src = values["source"] as? Int,
               let lat = values["latitude"] as? Double,
               let lon = values["longitude"] as? Double {
                setLocation(source: src, latitude: lat, longitude: lon)
                result(nil)
            } else {
              result(
                FlutterError(
                  code: Constants.kOpenWrapPlatformException,
                  message: "Error while calling setLocation on OpenWrapSDK class.",
                  details: "Cannot set location as the provided location data is invalid."
                )
              )
            }
            
        case "setCoppa":
            if let enabled = call.arguments as? Bool {
                OpenWrapSDK.setCoppaEnabled(enabled)
            }
            result(nil)
        case "setSSLEnabled":
            if let enabled = call.arguments as? Bool {
                OpenWrapSDK.setSSLEnabled(enabled)
            }
            result(nil)
        case "allowAdvertisingId":
            if let allow = call.arguments as? Bool {
                OpenWrapSDK.allowAdvertisingId(allow)
            }
            result(nil)
        case "setApplicationInfo":
            if let appInfo = call.arguments as? Dictionary<String, Any> {
                setApplicationInfo(applicationInfo: appInfo)
            }
            result(nil)
        case "setUserInfo":
            if let userInfo = call.arguments as? Dictionary<String, Any> {
                setUserInfo(userInfo: userInfo)
            }
            result(nil)
        default:
            /// Unsupported method call; error out
            result(FlutterMethodNotImplemented)
        }
    }
        
    static func setLocation(source: Int, latitude: Double, longitude:Double) {
        let location = CLLocation(latitude: latitude, longitude: longitude)
        let src = POBLocSource(rawValue: source + 1) ?? .userProvided
        OpenWrapSDK.setLocation(location, source: src)
    }

    static func setApplicationInfo(applicationInfo: Dictionary<String, Any>) {
        let app = POBApplicationInfo()
        if let url = applicationInfo["storeURL"] as? String {
            app.storeURL = URL(string: url) ?? app.storeURL
        }
        if let domain = applicationInfo["domain"] as? String {
            app.domain = domain
        }
        if let paid = applicationInfo["paid"] as? Bool {
            app.paid = paid ? POBBOOLYes : POBBOOLNo
        }
        if let cat = applicationInfo["categories"] as? String {
            app.categories = cat
        }
        if let keywords = applicationInfo["appKeywords"] as? String {
            app.keywords = keywords
        }
        OpenWrapSDK.setApplicationInfo(app)
    }

    static func setUserInfo(userInfo:Dictionary<String, Any>) {
        let uInfo = POBUserInfo()
        if let genderValue = userInfo["gender"] as? Int {
          switch genderValue {
            case 0:
              uInfo.gender = .male

            case 1:
              uInfo.gender = .female

            case 2:
              uInfo.gender = .other

            default:
              uInfo.gender = .other
          }
        }
        if let yob = userInfo["birthYear"] as? NSNumber {
            uInfo.birthYear = yob
        }
        if let metro = userInfo["metro"] as? String {
            uInfo.metro = metro
        }
        if let city = userInfo["city"] as? String {
            uInfo.city = city
        }
        if let region = userInfo["region"] as? String {
            uInfo.region = region
        }
        if let country = userInfo["country"] as? String {
            uInfo.country = country
        }
        if let zip = userInfo["zip"] as? String {
            uInfo.zip = zip
        }
        if let keywords = userInfo["userKeywords"] as? String {
            uInfo.keywords = keywords
        }
        OpenWrapSDK.setUserInfo(uInfo)
    }
}
